File: src/main/kotlin/easy/TwoSum.kt
package easy

class TwoSum {
    /*
    * Link: https://leetcode.com/problems/two-sum/
    * Difficulty: Easy
    */
    fun twoSum(nums: IntArray, target: Int): IntArray {
        val hashMap = mutableMapOf<Int, Int>()

        // This is a linear solution with O(N) space complexity
        nums.forEachIndexed { index, value ->
            val diff = target - value

            if(hashMap.contains(diff)){
                return@twoSum intArrayOf(hashMap[diff]!!, index)
            }

            hashMap[value] = index
        }
        throw Throwable("This problem has no solution!")
    }
}

File: src/easy/two_sum.go
package easy

/*
Difficult: Easy
Link: https://leetcode.com/problems/two-sum/
*/
func twoSum(nums []int, target int) []int {
	mapping := make(map[int]int)

	for i, n := range nums {
		if index, ok := mapping[target-n]; ok {
			return []int{i, index}
		}
		mapping[n] = i
	}

	return nil
}
